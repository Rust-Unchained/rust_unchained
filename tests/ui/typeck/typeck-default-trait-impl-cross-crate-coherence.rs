//@ aux-build:tdticc_coherence_lib.rs

#![feature(negative_impls)]

extern crate tdticc_coherence_lib as lib;

use lib::DefaultedTrait;

struct A;

impl DefaultedTrait for (A,) {} //~ ERROR E0117

struct B;

impl !DefaultedTrait for (B,) {} //~ ERROR E0117

struct C;
struct D<T>(T);

impl DefaultedTrait for Box<C> {} //~ ERROR E0321

impl DefaultedTrait for lib::Something<C> {} //~ ERROR E0321

impl DefaultedTrait for D<C> {} // OK in standard Rust

fn main() {}
