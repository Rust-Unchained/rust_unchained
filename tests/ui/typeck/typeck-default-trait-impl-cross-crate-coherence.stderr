error[E0321]: cross-crate traits with a default impl, like `DefaultedTrait`, can only be implemented for a struct/enum type, not `(A,)`
  --> $DIR/typeck-default-trait-impl-cross-crate-coherence.rs:11:1
   |
LL | impl DefaultedTrait for (A,) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't implement cross-crate trait with a default impl for non-struct/enum type

error[E0321]: cross-crate traits with a default impl, like `DefaultedTrait`, can only be implemented for a struct/enum type, not `(B,)`
  --> $DIR/typeck-default-trait-impl-cross-crate-coherence.rs:15:1
   |
LL | impl !DefaultedTrait for (B,) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't implement cross-crate trait with a default impl for non-struct/enum type

error[E0321]: cross-crate traits with a default impl, like `DefaultedTrait`, can only be implemented for a struct/enum type defined in the current crate
  --> $DIR/typeck-default-trait-impl-cross-crate-coherence.rs:20:1
   |
LL | impl DefaultedTrait for Box<C> {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't implement cross-crate trait for type in another crate

error[E0321]: cross-crate traits with a default impl, like `DefaultedTrait`, can only be implemented for a struct/enum type defined in the current crate
  --> $DIR/typeck-default-trait-impl-cross-crate-coherence.rs:22:1
   |
LL | impl DefaultedTrait for lib::Something<C> {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't implement cross-crate trait for type in another crate

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0321`.
